lib/awful/completion.lua

```diff
@@ -154,10 +161,25 @@ function completion.shell(command, cur_pos, ncomp, shell)
         while true do
             local line = c:read("*line")
             if not line then break end
-            if os.execute("test -d " .. string.format('%q', line)) == 0 then
-                line = line .. "/"
+            if comptype == "command" then
+                if str_starts(line, ".") or str_starts(line, "/") then
+                    if gfs.file_executable(line) then
+                        table.insert(output, bash_escape(line))
+                    end
+                else
+                    for path in string.gmatch(os.getenv('PATH'), "[^:]+") do
+                        if gfs.file_executable(path .. "/" .. line) then
+                            table.insert(output, bash_escape(line))
+                            break
+                        end
+                    end
+                end
+            end
+            if gfs.is_dir(line) then
+                table.insert(output, bash_escape(line .. "/"))
+            elseif comptype == "file" then
+                table.insert(output, bash_escape(line))
             end
-            table.insert(output, bash_escape(line))
```

Comment:
Yes, this sounds reasonable.

Do you want the trailing slash on directories? Then we need a check for `starts_with "./"` and `is_dir`, but it's way shorter than my current construct.

I will push these changes as separate commits to keep this discussion, but I can squash it after review.